tools/user_migration/migrate_user.py

```diff
@@ -0,0 +1,288 @@
+#! venv/bin/python
+
+# painfully stupid script to check out dumping mysql databases to avro.
+# Avro is basically "JSON" for databases. It's not super complicated & it has
+# issues (one of which is that it requires Python2).
+#
+#
+
+import argparse
+import base64
+import time
+from datetime import datetime
+
+from mysql import connector
+from google.cloud import spanner
+from urllib.parse import urlparse
+
+SPANNER_NODE_ID = 800
+
+
+class BadDSNException(Exception):
+    pass
+
+
+def get_args():
+    parser = argparse.ArgumentParser(
+        description="move user from sql to spanner")
+    parser.add_argument(
+        '--dsns', default="move_dsns.lst",
+        help="file of new line separated DSNs")
+    parser.add_argument(
+        '--users', default="move_users.lst",
+        help="file of new line separated users to move")
+    parser.add_argument(
+        '--token_dsn',
+        help="DSN to the token server database (optional)"
+    )
+    parser.add_argument(
+        '--verbose',
+        action="store_true",
+        help="verbose logging"
+    )
+    return parser.parse_args()
+
+
+def conf_mysql(dsn):
+    """create a connection to the original storage system """
+    connection = connector.connect(
+        user=dsn.username,
+        password=dsn.password,
+        host=dsn.hostname,
+        port=dsn.port or 3306,
+        database=dsn.path[1:]
+    )
+    return connection
+
+
+def conf_spanner(dsn):
+    """create a connection to the new Spanner system"""
+    path = dsn.path.split("/")
+    instance_id = path[-3]
+    database_id = path[-1]
+    client = spanner.Client()
+    instance = client.instance(instance_id)
+    database = instance.database(database_id)
+    return database
+
+
+def conf_db(dsn):
+    """read the list of storage definitions from the file and create
+    a set of connetions.
+
+     """
+    if dsn.scheme == "mysql":
+        return conf_mysql(dsn)
+    if dsn.scheme == "spanner":
+        return conf_spanner(dsn)
+    raise RuntimeError("Unknown DNS type: {}".format(dsn.scheme))
+
+
+def update_token(databases, user):
+    """optionally update the TokenServer storage indicating the user
+    is now on Spanner
+
+    """
+    cursor = databases['token'].cursor()
+    cursor.execute(
+        """
+        UPDATE
+            users
+        SET
+            replaced_at = {timestamp},
+            nodeid = {nodeid}
+        WHERE
+           uid = {uid}
+        """.format(
```

Comment:
I can't help but be worried by regular str interpolation vs bind parameters but I guess it's harmless here

Reply:
Not a fan either, but I couldn't find an interpolation function inside of `mysql.connector`. The examples in the library use python string formatters. Since this is being executed by trusted individuals, I'm not quite as uneasy.